lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -1806,68 +1802,202 @@ func networkFromContainerInfo(vc *viccontainer.VicContainer, info *models.Contai
+	// will contain a combined set of port mappings
+	if portMap == nil {
+		portMap = make(nat.PortMap)
+	}
 
-	// Iterate over the hostconfig that was set in docker create.  Get non-nil
-	// bindings and fix up ip addr and add to networks
-	if vc.HostConfig != nil && vc.HostConfig.PortBindings != nil {
-		//		networks.Ports = vc.HostConfig.PortBindings
-		for port, portbindings := range vc.HostConfig.PortBindings {
+	// add IP address into port spec to allow direct usage of data returned by calls such as docker port
+	var ip string
+	ips, _ := publicIPv4Addrs()
+	if len(ips) > 0 {
+		ip = ips[0]
+	}
 
-			var newbindings []nat.PortBinding
+	// Preserve the existing behaviour if we do not have an IP for some reason.
+	if ip == "" {
+		ip = "0.0.0.0"
+	}
 
-			for _, binding := range portbindings {
-				nb := binding
+	for _, ep := range endpoints {
+		if ep.Direct {
+			continue
+		}
 
-				// Check host IP.  VIC only support 0.0.0.0
-				if nb.HostIP == "" {
-					nb.HostIP = "0.0.0.0"
+		for _, port := range ep.Ports {
+			mappings, err := nat.ParsePortSpec(port)
+			if err != nil {
+				log.Error(err)
+				// just continue if we do have partial port data
+			}
+
+			for i := range mappings {
+				p := mappings[i].Port
+				b := mappings[i].Binding
+
+				if b.HostIP == "" {
+					b.HostIP = ip
 				}
 
-				newbindings = append(newbindings, nb)
-			}
+				if mappings[i].Binding.HostPort == "" {
+					// leave this undefined for dynamic assignment
+					// TODO: for port stability over VCH restart we would expect to set the dynamically assigned port
+					// recorded in containerVM annotations here, so that the old host->port mapping is preserved.
+				}
+
+				log.Debugf("Adding indirect mapping for port %v: %v (%s)", p, b, port)
 
-			portMap[port] = newbindings
+				current, _ := portMap[p]
+				portMap[p] = append(current, b)
+			}
 		}
 	}
 
-	// Iterate over the container's original image config.  This is the set of
-	// exposed ports.  For ports that were not in hostConfig, we assign value of
-	// nil.  This appears to be the behavior of regular docker.
-	if vc.Config != nil {
-		for port := range vc.Config.ExposedPorts {
-			if _, ok := portMap[port]; ok {
-				continue
+	return portMap
+}
+
+// addDirectEndpointsToPortMap() constructs a docker portmap from the container's info as returned by the portlayer for those
+// ports exposed directly from the containerVM via container network
+// The portMap provided is modified and returned - the return value should always be used.
+func addDirectEndpointsToPortMap(endpoints []*models.EndpointConfig, portMap nat.PortMap) nat.PortMap {
+	if len(endpoints) == 0 {
+		return portMap
+	}
+
+	if portMap == nil {
+		portMap = make(nat.PortMap)
+	}
+
+	for _, ep := range endpoints {
+		if !ep.Direct {
+			continue
+		}
+
+		// add IP address into the port spec to allow direct usage of data returned by calls such as docker port
+		var ip string
+		rawip, _, _ := net.ParseCIDR(ep.Address)
```

Comment:
s/rawip/rawIP

We should also add a `#nosec` comment here since we're ignoring the returned error.